fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/BulkDataUtils.java

```diff
@@ -155,41 +156,89 @@ private static int getFhirResourceFromBufferReader(BufferedReader resReader, int
+                // Notify s3 client to abort and prevent the server from keeping on sending data.
+                s3InStream.abort();
+            } catch (Exception ioe) {
+                logger.warning("readFhirResourceFromObjectStore: " + "Error proccesing file " + itemName + " - " + ioe.getMessage());
+                exported = 0;
+            }
         }
+
         return exported;
     }
 
 
-    public static int readFhirResourceFromLocalFile(String filePath, int numOfLinesToSkip, List<Resource> fhirResources) {
-        int exported;
-        try (BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath))) {
-            exported = getFhirResourceFromBufferReader(resReader, numOfLinesToSkip, fhirResources);
-        } catch (Exception ioe) {
-            logger.warning("readFhirResourceFromLocalFile: " + "Error proccesing file " + filePath + " - " + ioe.getMessage());
-            exported = 0;
+    public static int readFhirResourceFromLocalFile(String filePath, int numOfLinesToSkip, List<Resource> fhirResources,
+            boolean isReuseInput, ImportTransientUserData transientUserData) {
+        int exported = 0;
+        if (isReuseInput) {
+            try {
+                if (transientUserData.getBufferReader() == null) {
+                    BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath));
+                    transientUserData.setBufferReader(resReader);
+                }
+                exported = getFhirResourceFromBufferReader(transientUserData.getBufferReader(), 0, fhirResources);
+            } catch (Exception ioe) {
+                logger.warning("readFhirResourceFromLocalFile: " + "Error proccesing file " + filePath + " - " + ioe.getMessage());
+                exported = 0;
+            }
+        } else {
+            try (BufferedReader resReader = Files.newBufferedReader(Paths.get(filePath))) {
+                exported = getFhirResourceFromBufferReader(resReader, numOfLinesToSkip, fhirResources);
+            } catch (Exception ioe) {
+                logger.warning("readFhirResourceFromLocalFile: " + "Error proccesing file " + filePath + " - " + ioe.getMessage());
+                exported = 0;
+            }
         }
         return exported;
     }
 
 
-    public static int readFhirResourceFromHttps(String dataUrl, int numOfLinesToSkip, List<Resource> fhirResources) {
-        int exported;
-        try (BufferedReader resReader = new BufferedReader(new InputStreamReader(new URL(dataUrl).openConnection().getInputStream()))) {
-            exported = getFhirResourceFromBufferReader(resReader, numOfLinesToSkip, fhirResources);
-        } catch (Exception ioe) {
-            logger.warning("readFhirResourceFromHttps: " + "Error proccesing file " + dataUrl + " - " + ioe.getMessage());
-            exported = 0;
+    public static int readFhirResourceFromHttps(String dataUrl, int numOfLinesToSkip, List<Resource> fhirResources,
```

Comment:
I think there should be some defensive programming here to validate the url is acceptable.  This may be best at the REST layer.

Reply:
agreed, make sense for me.